### DISCUSSION TITLE
Migrating Homebrew install from one Macbook to another (both Intel) with `brew bundle`: what's a Brewfile, and how to migrate from source filesystem (of a dead MacBook)?

### DISCUSSION THREAD

**Discussion by johnnyutahh** [Homebrew/discussions#3664] [Everyday usage] [answered]
I want to minmally/efficiently understand macOS `/usr/local` (presumably where _all_ of Homebrew downloads/builds/etc live... yes?) filesystem layout/structure/etc in order to migrate a `/usr/local`  from one Macbook to another, both running macOS v10.15.7, but with different hardware configurations (both running Intel-based CPUs/cores)... **IF** a migration (or a `/usr/local` merge of some kind?) is feasible.

**QUESTION**: can someone provide guidance for more _efficient_ understanding/learning (for the above purpose)?

Something other than `man brew` (which appears to go deeper than I need to understand the migration maneuver)?

**My background**

I am a long-time (over a decade) macOS-Homebrew user. (I've also dabbled a little with Linux-based Homebrew installs as well.) I know enough to install all sorts of brew-based software (including running services like my own DNS server on macOS), but I'm NOT familiar with the guts of Homebrew. Pls do not treat me as a first-time user; rather please consider me a first time "Homebrew 'guts' evaluator"... or something like that.

I certainly have never written a Homebrew software package/formula/etc. I _have_ developed many software packages across many other OS platforms (numerous Unix-es, Linux, Windows, etc), so I am far from "new" to software development and deployment... except on macOS and Homebrew (both).

**Related discussion**

https://www.reddit.com/r/MacOS/comments/xh59gd/trying_to_learn_homebrewmacos_usrlocal_layout_for/

**Comment by carlocab** [comment]
This doesn't answer your question about the structure of directories inside `HOMEBREW_PREFIX`, but perhaps you'd like to consider using `brew bundle` to migrate your Homebrew installation from one Mac to another.

Using `brew bundle` does not require any knowledge of the aforementioned structure.

**Comment by johnnyutahh** [comment]
`brew bundle` looks promising, thank you!  After skimming https://github.com/Homebrew/homebrew-bundle and `brew help bundle` the first, key question I have is:  WHAT is a Brewfile?  There's lots of docs/talk about _a_ Brewfile or Brewfiles... but no (at least thus far, for me?) easy-to-find place/discussion about WHAT a Brewfile actually _is_.

Any help for the above is greatly appreciated.  :-)

My first (total) guess: a Brewfile is something akin to a `.tar` file... but this again is a total guess.

**Reply by carlocab** [reply]
There is en example `Brewfile` shown in the `README` at https://github.com/Homebrew/homebrew-bundle.

**Reply by johnnyutahh** [reply]
> There is en example `Brewfile` shown in the `README` at https://github.com/Homebrew/homebrew-bundle.

Yes... I saw that. But that does not answer my question. I'm seeking to understand what the purpose/context/etc of a Brewfile is. Is it an "input" or an "output" (or both) to `brew bundle [something]` command?  Is it "config-ish" file that a human edits? Is it a perpetual config-ish file (that is often found in `/etc` or `/usr/local/etc`)? Maybe it's something else?

Is there something I'm clearly missing?  I sincerely would very much like to be shown something obvious to tell me I'm "wrong" and that I'm simply not reading the docs correctly.

**Reply by carlocab** [reply]
> Yes... I saw that.

Sorry, I thought you hadn't, given this comment:

> My first (total) guess: a Brewfile is something akin to a `.tar` file... but this again is a total guess.

In any case, you don't need to know what a `Brewfile` is. Just do
```
brew bundle dump
```
then copy the `Brewfile` in your current working directory to your new machine.

After installing Homebrew on the new machine, do
```
brew bundle --file=/path/to/Brewfile
```

**Reply by johnnyutahh** [reply]
> In any case, you don't need to know what a `Brewfile` is. Just do...

Excellent, thank you!  Paraphrashing the procedure:

1. on source-macOS machine: `brew bundle dump`  --> writes `Brewfile` to current directory.
2. on target-macOS machine`brew bundle --file=/path/to/Brewfile_from_source_macOS_machine` --> re-installs Homebrew environment

Based upon the above procedure: here's my current interpretation of what a Brewfile is, outlining the dump-and-rebuild process of `brew bundle` along the way (one can make a case that adding a paragraph like this to the help/man docs would go a long way):

_"The Brewfile [apparently?] is an auto-generated (by `brew bundle dump`) text-config-file that specifies the entire Homebrew-installed "environment" [but apparently not `/usr/local/etc` config files, according to my initial tests] and is used as input to `brew bundle --file=[input_Brewfile]` to rebuild said environment on a new machine. The Brewfile tends to be kilobytes in size and not megabytes big."_

**Comment by johnnyutahh** [comment]
Regardless of the Brewfile debate:

I need a procedure to migrate the `HOMEBREW_PREFIX` (often `/usr/local`... I guess?) from one macOS system to another. Apparently `brew bundle` is designed to provide the core means to do this, BUT I do not know any example procedure for doing this. Hence to me, understanding what the heck a Brewfile is supposed to be (it's purpose, etc) would go a long way for me to figure out he procedure on my own.

Alas, for now: it seems like I have to "reverse engineer" this understanding from the semi-cryptic (to me, thus far) docs.

**Reply by SMillerDev** [reply]
> I need a procedure to migrate the HOMEBREW_PREFIX (often /usr/local... I guess?) from one macOS system to another.

There isn't any, since the differences in device could mean that nothing would work.

> Apparently brew bundle is designed to provide the core means to do this, BUT I do not know any example procedure for doing this. 

Example:
`brew bundle dump` on device A, copy the package definition that is a `Brewfile` to device B, run `brew bundle --file=Brefile` on device B to install everything again.

`man brew` and https://github.com/Homebrew/homebrew-bundle have more information about `brew bundle`

**Comment by johnnyutahh** [comment]
**_Next question_**:

How do I run a `brew bundle dump` on only the _filesystem_ (or `/usr/local` copy--if that's all that's needed?) of the original machine?

The original/source MacBook is not currently operational--it died--although target-disk-mode on the "dead" MacBook still works... thus far, but I do not want to count on it. I also have (at least) a full copy of the original-MacBook's `/usr/local` on my new MacBook and a full filesystem copy of the original MacBook on an external disk.

Reminder: both source and target MacBook's are Intel-based.

**Reply by carlocab** [reply]
Try
```sh
/path/to/dead/usr/local/bin/brew bundle dump
```

**Reply by johnnyutahh** [reply]
> Try
> 
> ```shell
> /path/to/dead/usr/local/bin/brew bundle dump
> ```

I tried this, and it (very unfortunately) does not appear to work, at least for my first few tests. It only generates a `Brewfile` for the current macOS/MacBook (which is the target MacBook with a `/path/to/dead/usr/local/` copy, of course).

Nice try, however -- thank you @carlocab!

Does anyone else have additional suggestions?

**Reply by johnnyutahh** [reply]
At this point, these seem like my only options; options 1, 2, and 3 listed below have no (current) solution, 4 is something I want to avoid, and 5 is the only feasible thing I've seen yet (but not yet tried):

1. figure out how to run `brew bundle dump` on a filesystem copy of the source/original macOS. (Have not yet found a way to do this.)
1. I figure out how to resuscitate my original (broken) MacBook and run the above procedure.
1. Figure out a way to run a `brew list` on the cloned `/usr/local` ( https://superuser.com/a/1141414/98033 ) -- which I'm not sure is possible.
1. Re-install everything by hand, manually, discovering every package I'm missing in a completely-ad-hoc/manual fashion. (I definitely want to _not_ choose this option, if I can get something else to work.)
1. `chroot`-to-source-disk-MacBook solution [update: does not appear to work because of `chroot`-on-macOS problems, see below]: https://www.reddit.com/r/MacOS/comments/xhih3d/comment/ioxu7im/?utm_source=reddit&utm_medium=web2x&context=3

**Reply by johnnyutahh** [reply]
Update: the following `chroot`-based solution is the most-feasible thing I've seen yet (and I'm editing my above comment to include this as listed item number 5, above):

https://www.reddit.com/r/MacOS/comments/xhih3d/comment/ioxu7im/?utm_source=reddit&utm_medium=web2x&context=3

**Reply by carlocab** [reply]
If you just need a list of formulae you had installed previously:
```
ls /path/to/dead/usr/local/Cellar
```
Though if this is a very old Homebrew install, then this list may include formulae that no longer exist. Additionally, if some of these formulae are from external taps, you'll need to fully specify the name (e.g. `brew install user/tap/formula-name`) to install them.

**Reply by johnnyutahh** [reply]
> If you just need a list of formulae you had installed previously:
> 
> ```
> ls /path/to/dead/usr/local/Cellar
> ```

Copy that, thanks. Another workaround, even though there would seem to be lots of environment-configuration settings not captured with this procedure... but again, it's better than nothing. Thanks again!

**Reply by johnnyutahh** [reply]
Update: `chroot` on macOS does not appear to work, and is known to be quite problematic, eg:

1. https://gist.github.com/eddieh/b3c04436eb6b28826939cd57f1b704bb
1. https://apple.stackexchange.com/questions/356970/how-do-i-create-a-minimal-chroot-jail-in-macos

I'm back to having no realistic, automatic option to capture the entire Homebrew environment from the source/original macOS/MacBook... unless I can revive the broken (source/original) MacBook.  :-/

My system:

```
$ sw_vers
ProductName:	Mac OS X
ProductVersion:	10.15.7
BuildVersion:	19H15
$ which chroot
/usr/sbin/chroot
$ sudo chroot /Volumes/external_disk
Killed: 9
$ pwd
/Users/myuser
$ source ~/.bash_profile
$ which chroot
/usr/local/opt/coreutils/libexec/gnubin/chroot
$ sudo chroot /Volumes/external_disk
Killed: 9
$ pwd
/Users/myuser
$ chroot --version
chroot (GNU coreutils) 9.1
Copyright (C) 2022 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <https://gnu.org/licenses/gpl.html>.
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.

Written by Roland McGrath.
$
```

**Reply by johnnyutahh** [reply]
Update:

a.`mv /usr/local` to a temporary directory in order to link-to/replace with the copied-from-original/source-MacBook `/usr/local` also appears not to work. `sudo mv /usr/local` blocks me.

b. what if I just delete the current `/usr/local` on the target/new MacBook and try to move the old `/usr/local` there... might that possibly work??

**Reply by johnnyutahh** [reply]
> Update: `chroot` on macOS does not appear to work, and is known to be quite problematic...

fyi. This may have been due to a broken environment in/on/within my macOS; I'm not yet convinced that `chroot` can not work for a similar case in the future.

**Reply by johnnyutahh** [reply]
> b. what if I just [first uninstall Homebrew then] delete the current `/usr/local` on the target/new MacBook and try to move the old `/usr/local` there [and then reinstall Homebrew and run `brew bundle dump` to generate the original/source brew environment\]... might that possibly work??

SOLUTION: the above is exactly what I ended up doing.

(Whew... finally got to "the end". 😎)

**Reply by johnnyutahh** [reply]
How do I mark the above "discussion thread leaf" as `The Answer`?

**Comment by johnnyutahh** [comment]
My current perspective: Homebew's inability to run-time manage a non-default-install directory is a major problem for this scenario. It's also arguable that it's not unreasonable for Homebrew to have (some time ago) considered install/file-location modularity (at run time) in its architecture/design, especially given Homebrew's maturity. Alas, mine is a subjective argument.

In general, I'm surprised there (appears to be?) no `brew --path-prefix=[non_default_path_prefix]`-like feature currently in Homebrew. :-/  I'd be more surprised if there's not a feature-request/roadmap with this (feature) on the books.

Nonetheless, Homebrew has provided me and my team ENORMOUS (and hard-to-measure) value. Pls do not let my small grip here distract from my HUGE appreciation of the project. :-)

**Comment by johnnyutahh** [comment, marked as answer]
Here's my answer for this problem (a post I made elsewhere in this discussion thread), for today anyway:

https://github.com/orgs/Homebrew/discussions/3664#discussioncomment-3676106
